crates/typst-pdf/src/color.rs

```diff
@@ -42,106 +41,168 @@ impl ColorSpaces {
                 .range([0.0, 1.0, 0.0, 1.0, 0.0, 1.0])
                 .filter(Filter::FlateDecode);
         }
 
         // Write the sRGB color space.
-        if let Some(srgb) = self.srgb {
+        if self.use_srgb {
             chunk
-                .icc_profile(srgb, &SRGB_ICC_DEFLATED)
+                .icc_profile(refs.srgb, &SRGB_ICC_DEFLATED)
                 .n(3)
                 .range([0.0, 1.0, 0.0, 1.0, 0.0, 1.0])
                 .filter(Filter::FlateDecode);
         }
 
         // Write the gray color space.
-        if let Some(gray) = self.d65_gray {
+        if self.use_d65_gray {
             chunk
-                .icc_profile(gray, &GRAY_ICC_DEFLATED)
+                .icc_profile(refs.d65_gray, &GRAY_ICC_DEFLATED)
                 .n(1)
                 .range([0.0, 1.0])
                 .filter(Filter::FlateDecode);
         }
     }
+
+    pub fn merge(&mut self, other: &Self) {
+        self.use_d65_gray |= other.use_d65_gray;
+        self.use_linear_rgb |= other.use_linear_rgb;
+        self.use_oklab |= other.use_oklab;
+        self.use_srgb |= other.use_srgb;
+    }
+}
+
+/// Write the color space.
+pub fn write(
+    color_space: ColorSpace,
+    writer: writers::ColorSpace,
+    refs: &ColorFunctionRefs,
+) {
+    match color_space {
+        ColorSpace::Oklab | ColorSpace::Hsl | ColorSpace::Hsv => {
+            let mut oklab = writer.device_n([OKLAB_L, OKLAB_A, OKLAB_B]);
+            write(ColorSpace::LinearRgb, oklab.alternate_color_space(), refs);
+            oklab.tint_ref(refs.oklab);
+            oklab.attrs().subtype(DeviceNSubtype::DeviceN);
+        }
+        ColorSpace::Oklch => write(ColorSpace::Oklab, writer, refs),
+        ColorSpace::Srgb => writer.icc_based(refs.srgb),
+        ColorSpace::D65Gray => writer.icc_based(refs.d65_gray),
+        ColorSpace::LinearRgb => {
+            writer.cal_rgb(
+                [0.9505, 1.0, 1.0888],
+                None,
+                Some([1.0, 1.0, 1.0]),
+                Some([
+                    0.4124, 0.2126, 0.0193, 0.3576, 0.715, 0.1192, 0.1805, 0.0722, 0.9505,
+                ]),
+            );
+        }
+        ColorSpace::Cmyk => writer.device_cmyk(),
+    }
+}
+
+pub struct ColorFunctionRefs {
+    oklab: Ref,
+    srgb: Ref,
+    d65_gray: Ref,
+}
+
+impl Default for ColorFunctionRefs {
+    fn default() -> Self {
+        Self {
+            oklab: Ref::new(1),
+            srgb: Ref::new(1),
+            d65_gray: Ref::new(1),
+        }
+    }
+}
+
+impl Renumber for ColorFunctionRefs {
+    fn renumber(&mut self, old: Ref, new: Ref) {
+        self.oklab.renumber(old, new);
+        self.srgb.renumber(old, new);
+        self.d65_gray.renumber(old, new);
+    }
+}
+
+pub fn alloc_color_functions_refs(
+    _context: &AllocGlobalRefs,
+) -> (PdfChunk, ColorFunctionRefs) {
+    let mut chunk = PdfChunk::new();
+
+    let refs = ColorFunctionRefs {
+        oklab: chunk.alloc(),
```

Comment:
we allocate all three references here always, even if they are not used, right? so the xref table will just be empty for those?